examples/server/server.cpp

```diff
@@ -0,0 +1,794 @@
+        server_print_usage(argc, argv, default_params);
+        exit(1);
+  }
+
+  if (params.seed <= 0)
+  {
+    params.seed = time(NULL);
+  }
+
+  fprintf(stderr, "%s: seed = %d\n", __func__, params.seed);
+
+  Llama* llama = new Llama(params);
+  if(!llama->load_context()) {
+    return 1;
+  }
+
+  Server svr;
+
+  svr.Get("/", [](const Request &req, Response &res)
+          {
+            res.set_content("<h1>llama.cpp server works</h1>", "text/html");
+          }
+  );
+
+  svr.Post("/setting-context", [&llama](const Request &req, Response &res) {
+            if(!llama->context_config) {
+              json body = json::parse(req.body);
+              /*
+                Seed whould be passed by the request, but seem
+                the current implementation need it in the load file
+              */
+              if (!body["threads"].is_null())
+              {
+                llama->params.n_threads = body["threads"].get<int>();
+              }
+              if (!body["n_predict"].is_null())
+              {
+                llama->params.n_predict = body["n_predict"].get<int>();
+              }
+              if (!body["top_k"].is_null())
+              {
+                llama->params.top_k = body["top_k"].get<int>();
+              }
+              if (!body["top_p"].is_null())
+              {
+                llama->params.top_p = (float)body["top_p"].get<float>();
+              }
+              if (!body["temperature"].is_null())
+              {
+                llama->params.temp = (float)body["temperature"].get<float>();
+              }
+              if (!body["batch_size"].is_null())
+              {
+                llama->params.n_batch = body["batch_size"].get<int>();
+              }
+              if (!body["tags"].is_null())
+              {
+                json tags = body["tags"].get<json>();
+                llama->user_tag = tags["user"].get<std::string>();
+                llama->assistant_tag = tags["assistant"].get<std::string>();
+              }
+              if (!body["context"].is_null())
+              {
+                llama->params.prompt = "";
+                std::vector<json> context_messages = body["context"].get<std::vector<json>>();
+                for (json ctx_msg : context_messages)
+                {
+                  auto role = ctx_msg["role"].get<std::string>();
+                  if (role == "system")
+                  {
+                    llama->params.prompt = ctx_msg["content"].get<std::string>() + "\n\n";
+                  }
+                  else if (role == "user")
+                  {
+                    llama->params.prompt += llama->user_tag + " " + ctx_msg["content"].get<std::string>() + "\n";
+                  }
+                  else if (role == "assistant")
+                  {
+                    llama->params.prompt += llama->assistant_tag + " " + ctx_msg["content"].get<std::string>() + "\n";
+                  }
```

Comment:
This formatting seems too fixed, with the spaces and newlines. I don't think it can work for Alpaca at all.

What happens if the role is not `system` or `user` or `assistant`?

Reply:
Initially, this server implementation was more geared towards serving as a chatbot-like interaction. It was roughly based on the OpenAI [ChatGPT API](https://platform.openai.com/docs/guides/chat) although there were unfortunately many limitations such as the fact that prompt evaluation is a bit slow, so the initial context is constant from the beginning. It is also not possible to reset the context (without the need to reload the model) or at least I haven't found a way to do so.

Reply:
When `role` is not `user`, `assistant` or `system` the content is ignored.

Related Alpaca model, passing the following options in setting-context should work:

```javascript
const axios = require('axios');

async function Test() {
    let result = await axios.post("http://127.0.0.1:8080/setting-context", {
        context: [
            { role: "system", content: "Below is an instruction that describes a task. Write a response that appropriately completes the request." },
            { role: "user", content: "Write 3 random words" },
            { role: "assistant", content: "1. Sunshine\n2. Elephant\n3. Symphony" }
        ],
        tags: { user: "### Instruction:", assistant: "### Response:" },
        batch_size: 256,
        temperature: 0.2,
        top_k: 40,
        top_p: 0.9,
        n_predict: 2048,
        threads: 5
    });
    result = await axios.post("http://127.0.0.1:8080/set-message", {
        message: 'How to do a Hello word in C++, step by step'
    });
    if(result.data.can_inference) {
		result = await axios.get("http://127.0.0.1:8080/completion?stream=true", { responseType: 'stream' });
        result.data.on('data', (data) => {
            let dat = JSON.parse(data.toString());
            // token by token completion
            process.stdout.write(dat.content);
            if(dat.stop) {
                console.log("Completed");
            }
        });
    }
}

Test();
```

Output:
```bash
node test.js


1. Include the <iostream> header file
2. Use the "cout" object to output the text "Hello World!"
3. Add a new line character at the end of the output
4. Display the output on the screen

Example code:
'''c++
#include <iostream>

int main() {
    cout << "Hello World!" << endl;
    return 0;
}
'''
Completed
```

Reply:
I mean it will actually format it as `### Instruction: Write 3 random words\n### Response: 1. Sunshine\n2. Elephant\n3. Symphony` when [Alpaca is trained](https://github.com/tloen/alpaca-lora/blob/main/templates/alpaca.json) on `### Instruction:\nWrite 3 random words\n\n### Response:\n1. Sunshine\n2. Elephant\n3. Symphony`. It does work but that's because LLMs are flexible, but it's not optimal.

Don't get me wrong, I am trying to look out for you, if you decide to add this prompt templating into the API then you will be fixing and adding features to it for a long time when people come up with new models and ideas.

How about this?

```javascript
const context = [
    { user: "Write 3 random words", assistant: "1. Sunshine\n2. Elephant\n3. Symphony" }
]
const system = 'Below is an instruction that describes a task. Write a response that appropriately completes the request.'
const message = 'How to do a Hello word in C++, step by step'
const prompt = `${system}

${context.map(c => `### Instruction:
${c.user}

### Response:
${c.assistant}`)}

### Instruction:
${message}
`

// I think these two calls could be just merged into one endpoint
await axios.post("http://127.0.0.1:8080/setting-context", {
    content: prompt,
    batch_size: 256, // actually these could be command line arguments, since
    threads: 5,      // they are relevant to the server CPU usage
})
let result = await axios.get("http://127.0.0.1:8080/completion?stream=true", {
    responseType: 'stream', 
    temperature: 0.2,  // sampling parameters are used only for
    top_k: 40,         // generating new text
    top_p: 0.9,
    n_predict: 2048,
    
    // this is why it's nice to be able to get tokens
    keep: (await axios.post("http://127.0.0.1:8080/tokenize", { content: system })).length,
    // alternative is to pass keep as a string itself, but then you need to check if its tokens match what was given beforebegin
    
    stop: ['###', '\n'], // stop generating when any of these strings is generated,
                         // when streaming you have to make sure that '#' or '##' is not
                         // shown to the user because it could be a part of the stop keyword
                         // so there needs to be a little buffer.
})
let answer = ''
result.data.on('data', (data) => {
    let dat = JSON.parse(data.toString())
    // token by token completion
    process.stdout.write(dat.content)
    answer += dat.content
    if(dat.stop) {
        console.log("Completed")
        // save into chat log for the next question
        context.push({ user: message, assistant: answer })
    }
})
```

Actually there are libraries for this and other kinds of prompts in [langchain](https://js.langchain.com/docs/modules/prompts/prompt_templates/additional_functionality#few-shot-prompt-templates) for example.

Reply:
I have a problem with `cpp-httplib`. It seems that your current implementation only allows streaming of data with GET requests. You attempted to create a single 'completion' endpoint of type POST to allow for defining options, and it works. However, it doesn't work when you try to implement streaming, and you're unsure of the reason.

Reply:
I have a problem with cpp-httplib. And it's because it only allows streaming of data with GET requests. I tried to create a single POST endpoint called 'completion' to allow defining options, and it works. However, when it comes to streaming, it doesn't work for some reason.

Reply:
The idea you propose seems interesting, but there are limitations such as the time it takes to evaluate the prompt, especially if the prompts are very long. It also involves restarting the prompt, and I'm not sure how that could be done.

Reply:
Currently in this server, the `setting-context` needs to be called only once, `set-message` and `completion` need to be called whenever a response is desired, that allows for fast responses. The interaction context is being stored in "embd_inp" in the server instance.

I'm thinking of something like changing the behavior of the API with an option called `behavior` that has the choices `instruction`, `chat` and `generation`. There could also be an endpoint for performing embeddings and tokenization.

Reply:
If the client is sending a string that starts with the same text as last time, there is no need to evaluate it again, tokenize the new string, find how many tokens are the same at the start and set that as `n_past`. Anyway, at this point with a GPU we can evaluate even 512 tokens in less than 10 seconds (depending on hardware).

For the streaming, maybe the easiest solution is to have no streaming, just return the generated text. If the client wants to show individual tokens appearing to the user, it could just limit `n_predict` to 1 and call the API in a loop.

Reply:
I will try it. Good idea. I have to compare the new string tokens with `embd_inp`?